File: types/scenario.ts
import { z } from 'zod';

export const InitialLocationSchema = z.object({
    name: z.string().min(1, "Location name is required"),
    description: z.string().min(1, "Location description is required"),
});

export const CharacterConfigSchema = z.object({
    name: z.string().min(1, "Character name is required"),
    description: z.string().min(1, "Character description is required"),
    isPlayer: z.boolean(),
    initialLocationName: z.string().min(1, "Initial location is required"),
    encounterChance: z.number().min(0).max(1),
    goals: z.string().optional(),
});

export const ScenarioSchema = z.object({
    title: z.string().min(1, "Title is required"),
    description: z.string().min(1, "Description is required"),
    initialNarrativeTime: z.string().min(1, "Initial time is required"),
    locations: z.array(InitialLocationSchema).min(1, "At least one location is required"),
    characters: z.array(CharacterConfigSchema).min(1, "At least one character is required"),
    playerStartingLocation: z.string().min(1, "Player start location is required"),
});

export type ScenarioDefinition = z.infer<typeof ScenarioSchema>;

File: internal/models/scenario.go
package models

import "fmt"

// InitialLocation for scenario setup
type InitialLocation struct {
	Name        string `json:"name"`
	Description string `json:"description"`
}

// CharacterConfig for scenario setup
type CharacterConfig struct {
	Name                string  `json:"name"`
	Description         string  `json:"description"`
	IsPlayer            bool    `json:"isPlayer"`
	InitialLocationName string  `json:"initialLocationName"`
	EncounterChance     float64 `json:"encounterChance"`
	Goals               string  `json:"goals,omitempty"`
}

// ScenarioConfig for starting a new world
type ScenarioConfig struct {
	Title                 string            `json:"title"`
	Description           string            `json:"description"`
	InitialNarrativeTime  string            `json:"initialNarrativeTime"`
	Locations             []InitialLocation `json:"locations"`
	Characters            []CharacterConfig `json:"characters"`
	PlayerStartingLocation string           `json:"playerStartingLocation"`
}

// WorldState is the entire world state
type WorldState struct {
	ID                   string            `json:"id"`
	Scenario             ScenarioConfig    `json:"scenario"`
	Time                 WorldTime         `json:"time"`
	Characters           []Character       `json:"characters"`
	LocationClusters     []LocationCluster `json:"locationClusters"`
	Locations            []Location        `json:"locations"`
	Events               []WorldEvent      `json:"events"`
	Conversations        []Conversation    `json:"conversations"`
	PlayerCharacterID    string            `json:"playerCharacterId"`
	MainConversationID   string            `json:"mainConversationId"`
}

// Validate checks if the scenario config is valid
func (s *ScenarioConfig) Validate() error {
	if s.Title == "" {
		return fmt.Errorf("title is required")
	}
	if s.Description == "" {
		return fmt.Errorf("description is required")
	}
	if s.InitialNarrativeTime == "" {
		return fmt.Errorf("initial narrative time is required")
	}
	if len(s.Locations) == 0 {
		return fmt.Errorf("at least one location is required")
	}
	if len(s.Characters) == 0 {
		return fmt.Errorf("at least one character is required")
	}
	hasPlayer := false
	for _, c := range s.Characters {
		if c.IsPlayer {
			hasPlayer = true
			break
		}
	}
	if !hasPlayer {
		return fmt.Errorf("at least one character must be a player")
	}
	if s.PlayerStartingLocation == "" {
		return fmt.Errorf("player starting location is required")
	}
	return nil
}

// ChatMessage represents a message in the main chat
type ChatMessage struct {
	ID      string `json:"id"`
	Role    string `json:"role"` // "user" or "assistant"
	Content string `json:"content"`
}
